grr/server/grr_response_server/gui/api_plugins/metadata_test.py

```diff
@@ -0,0 +1,382 @@
+    expected["field_bytes"] = expected_bytes
+
+    expected_uint32 = {"type": "integer", "format": "uint32"}
+    expected["field_uint32"] = expected_uint32
+
+    expected_sfixed32 = {"type": "integer", "format": "int32"}
+    expected["field_sfixed32"] = expected_sfixed32
+
+    expected_sfixed64 = {"type": "integer", "format": "int64"}
+    expected["field_sfixed64"] = expected_sfixed64
+
+    expected_sint32 = {"type": "integer", "format": "int32"}
+    expected["field_sint32"] = expected_sint32
+
+    expected_sint64 = {"type": "integer", "format": "int64"}
+    expected["field_sint64"] = expected_sint64
+
+    expected_binarystream = {"type": "string", "format": "binary"}
+    expected["BinaryStream"] = expected_binarystream
+
+    self.assertDictEqual(expected, openapi_primitives)
+
+  def testRepetitiveField(self):
+    # Test the GET route. The repeated field should be a query parameter.
+    params = (self.openapi_desc_dict.get("paths").get("/metadata_test/method4")
+              .get("get").get("parameters"))
+    self.assertEqual(len(params), 1)
+
+    param = params[0]
+    self.assertEqual(param["in"], "query")
+    schema = param["schema"]
+    self.assertEqual(schema["type"], "array")
+    self.assertEqual(schema["items"]["type"], "integer")
+    self.assertEqual(schema["items"]["format"], "int64")
+
+    # Test the POST route. The repeated field should be in the request body.
+    req_body = (self.openapi_desc_dict.get("paths")
+                .get("/metadata_test/method4").get("post").get("requestBody"))
+    req_schema = req_body["content"]["application/json"]["schema"]
+    self.assertEqual(req_schema["type"], "object")
+    schema = req_schema["properties"]["field_repeated"]
+    self.assertEqual(schema["type"], "array")
+    self.assertEqual(schema["items"]["type"], "integer")
+    self.assertEqual(schema["items"]["format"], "int64")
+
+  def testEnumField(self):
+    # Test the GET route. The enum field should be a query parameter.
+    params = (self.openapi_desc_dict.get("paths").get("/metadata_test/method5")
+              .get("get").get("parameters"))
+    self.assertEqual(len(params), 1)
+
+    param = params[0]
+    self.assertEqual(param["in"], "query")
+    ref_get = param["schema"]["$ref"]
+    ref_nodes = ref_get.split("/")
+    ref_nodes = ref_nodes[1:]
+    schema = self.openapi_desc_dict
+    for node in ref_nodes:
+      schema = schema[node]
+
+    # Test the OpenAPI component agains its protobuf counterpart.
+    self.assertEqual(schema["type"], "integer")
+    self.assertEqual(schema["format"], "int32")
+    enum_values_openapi = set(schema["enum"])
+
+    message_d = MetadataEnumFieldMessage.protobuf.DESCRIPTOR
+    self.assertEqual(len(message_d.enum_types), 1)
+    enum_d = message_d.enum_types[0]
+    enum_values_ds = enum_d.values
+    enum_values_protobuf = set(
+      [enum_value_d.number for enum_value_d in enum_values_ds])
+    self.assertSetEqual(enum_values_openapi, enum_values_protobuf)
```

Comment:
I would use literals here instead of relying on reflection to fetch enum_values_protobuf. Given that we control the test data, we know what to expect and can explicitly list the enum values we expect.

Reply:
In general, in tests always prefer readability and obviousness over conciseness and being generic. Tests should read like a linear scenario rather then force the reader to do some mental execution of the code.

Reply:
Thank you both for explaining me this! I had limited experience writing tests, but I believe I've learned a lot by working on this and I'm sure there is more to learn about testing and software engineering&development in general. I've made the change to literals wherever I thought it possible, but I might have missed something. Please let me know if I should change something else.

Reply:
@alexandrucosminmihai - thanks for making the changes! I've made another review for the testing code, will resolve this comment so that we can continue with the latest review.